### PR TITLE
Feat: add compute network performance config block

### PR DESCRIPTION
**Pull request by brandonjbjelland**
## What is this change?

This PR adds the `network_performance_config` block to each of `resource_compute_instance`, `resource_compute_instance_from_template`, and `resource_compute_instance_template`. Resolves https://github.com/hashicorp/terraform-provider-google/issues/8735

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

TODO:

1. ~~This change SHOULD only apply toward the beta provider - I need to add the conditional rendering logic to make that happen.~~ - resolved
2. ~~My go isn't sharp enough to understand what's going on with the test case panicking here (I'm a humble baby gopher). A disk and image are successfully created but as the VM comes up, the test panics and I don't see why or know what steps to take to debug it. When this is resolved, I can absolutely go back and add similar tests to the other related resources. Any help here is appreciated~~ - resolved

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added a `network_performance_config` block to each of `resource_compute_instance`, `resource_compute_instance_from_template`, and `resource_compute_instance_template`
```

### CODE REVIEW

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown:205]: odd, this was unintentional and I'll fix it up.

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown:349]: Unrelated to my change but correcting the docs while in this file.

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:279]: Given the tight set of constraints for when this argument actually can take effect, I wasn't sure if we want to build any validation functions to fail fast if those conditions aren't met or if we're best to leave that to the API. I hedge on thinking the API should be responsible for those error states or input combinations that don't make sense.

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:497]: Unsure what has changed here.

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown:212]: This was unintentional and I'll fix it up in a follow up commit.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go:4467]: Can we use a smaller machine type here to avoid quota issues? I think our tests don't have anything higher than `n2-standard-2`

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go:1106]:
```suggestion
func TestAccComputeInstance_networkPerformanceConfig(t *testing.T) {
```

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go:1120]:
```suggestion
				Config: testAccComputeInstance_networkPerformanceConfig(imageName, diskName, instanceName),
```

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go:4444]:
```suggestion
func testAccComputeInstance_networkPerformanceConfig(disk string, image string, instance string) string {
```

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go:2676]: Crash is thrown here because `instance.NetworkPerformanceConfig` is returning nil. I'll address why in my top-level review comment

**Review by c2thorn — CHANGES_REQUESTED** [review]
@brandonjbjelland thanks for getting started on this! Unfortunately you've chosen to add a feature in one of our most complex resources, and there's a bit more effort needed to get this going.

You've added the fields to the schema correctly from what I can tell right now. Next is to edit the `resourceComputeInstanceCreate` and `resourceComputeInstanceRead` functions which do the actual writing/reading with the Compute client. This will lead you to create and use `expandNetworkPerformanceConfig` and `flattenNetworkPerformanceConfig` functions since you are adding a nested field. I think `guest_accelerator` is a similar field that you can use as an example.

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown
Comment (c2thorn) [reply mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown:349]: sounds good!

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (c2thorn) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:279]: I would agree. In addition, we've had similar validation functions become blockers in situations where the API has changed its own requirements.

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (c2thorn) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:497]: Probably a newline/whitespace.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go:4467]: I wish I could and actually I needed to raise quota in my project to test this also but the configuration only applies (from what I can tell) to machines with 32+ vCPUs. The [compatible types are on this docs page](https://cloud.google.com/compute/docs/networking/configure-vm-with-high-bandwidth-configuration#general_purpose_n2_machine_types).

I suppose if the configuration takes hold on a smaller instance type and has no practical egress bandwidth effect, would that be the preferred scenario to test?

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:279]: Thanks for the validation on my validation thought 😆

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go:2676]: Ahh many thanks! I'll have a peek and see if I can address this before tomorrow AM.

**Comment by brandonjbjelland** [comment]
@c2thorn - Thank you for taking a pass here and that makes sense. I'll try and flesh out the remaining bits you've described.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go
Comment (c2thorn) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go:4467]: Yes let's see if the API accepts the field on a smaller type and go with that if possible. If not, then yeah it looks like we'll have to raise quota on our test project. Let me know!

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:497]: Highlighting is my friend! This was whitespace at EOL 👍

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2749]: I assume not gating these behind the beta-only conditional is okay since they aren't a travelled code path in the GA provider.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (brandonjbjelland) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:4581]: This is what we settled on to test functionality here since `TIER_1` requires n2-x-32 or larger instances.

**Comment by brandonjbjelland** [comment]
I think I'm complete with my follow up here and downstream PRs are finally looking good. Please take another pass when possible @c2thorn .

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb
Comment (c2thorn) [inline mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb:296]: See my previous comment about switching to `Required: true`

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2716]: Can we place a nil check for `instance.NetworkPerformanceConfig` ?
If `instance.NetworkPerformanceConfig` is nil, we should error out before checking `instance.NetworkPerformanceConfig.TotalEgressBandwidthTier`. Just in case.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (c2thorn) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2723]: I think we still want to gate everything related to the field just for clarity. It makes the git history a bit cleaner and can provide some clarity for those browsing code in the GA provider itself.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (c2thorn) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:4555]: Thanks for getting this to work!

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (c2thorn) [inline mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:245]: Should the note about `GVNIC` apply to the instance template docs as well?

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (c2thorn) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2723]: This includes the other test function, instance helper functions, but not the HTML documentation. Documentation ends up getting merged for both versions.

**Review by c2thorn — CHANGES_REQUESTED** [review]
Looking very solid, just a few more minor asks

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2716]: Yes, sure thing.

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:2723]: Got it - I'll take care of that now.

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:245]: Yes I can add that. I was thinking `compute_instance`'s documentation was considered the full docs source where things get verbose for this family of resources but there's little cost to having the same text in both places.

File: mmv1/third_party/terraform/resources/resource_compute_instance.go.erb
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/resources/resource_compute_instance.go.erb:369]: Interesting, I hadn't noticed that pattern but it makes sense. Adding that now to both resources.

File: mmv1/third_party/terraform/resources/resource_compute_instance.go.erb
Comment (c2thorn) [inline mmv1/third_party/terraform/resources/resource_compute_instance.go.erb:369]: I wasn't clear earlier, the correct way to do this is to set
```suggestion
							Required:     true,
```

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/resources/resource_compute_instance.go.erb
Comment (c2thorn) [reply mmv1/third_party/terraform/resources/resource_compute_instance.go.erb]: Committing this since it's a one line change

**Comment by brandonjbjelland** [comment]
Thank you for your patience with me here. Ready for another pass here and hopefully we've arrived 🤞

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb
Comment (brandonjbjelland) [reply mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb]: Bleh - you were explicit here and I missed it. 🙏

**Comment by c2thorn** [comment]
Everything looks good from what I can tell! Gonna wait on our CI to pass checks + tests. If it all looks clear I'll merge!

**Review by c2thorn — APPROVED** [review]

**Pull request merged**
